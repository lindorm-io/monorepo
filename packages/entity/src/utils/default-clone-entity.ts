import { isFunction } from "@lindorm/is";
import { Constructor } from "@lindorm/types";
import { IEntity } from "../interfaces";
import { MetaColumnDecorator } from "../types";
import { globalEntityMetadata } from "./global";
import { parseColumn } from "./private";

const reset: Array<MetaColumnDecorator> = [
  "CreateDateColumn",
  "ExpiryDateColumn",
  "DeleteDateColumn",
  "UpdateDateColumn",
  "VersionColumn",
];

export const defaultCloneEntity = <E extends IEntity>(
  Entity: Constructor<E>,
  entity: E,
): E => {
  const metadata = globalEntityMetadata.get(Entity);
  const clone = new Entity();

  for (const column of metadata.columns) {
    if (metadata.generated.some((g) => g.key === column.key)) continue;

    if (
      reset.includes(column.decorator) ||
      metadata.primaryKeys.includes(column.key) ||
      metadata.generated.some((g) => g.key === column.key) ||
      metadata.indexes.some((u) => u.unique && u.keys.every((k) => k.key === column.key))
    ) {
      (clone as any)[column.key] = undefined;
    } else {
      (clone as any)[column.key] = parseColumn(column, entity);
    }

    if ((clone as any)[column.key]) continue;

    (clone as any)[column.key] = isFunction(column.fallback)
      ? column.fallback()
      : column.fallback;
  }

  const hooks = metadata.hooks.filter((h) => h.decorator === "OnCreate");

  for (const hook of hooks) {
    hook.callback(entity);
  }

  return clone;
};
