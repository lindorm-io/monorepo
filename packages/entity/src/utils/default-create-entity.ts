import { isFunction } from "@lindorm/is";
import { Constructor, DeepPartial } from "@lindorm/types";
import { IEntity } from "../interfaces";
import { globalEntityMetadata } from "./global";
import { parseColumn } from "./private";

export const defaultCreateEntity = <
  E extends IEntity,
  O extends DeepPartial<E> = DeepPartial<E>,
>(
  Entity: Constructor<E>,
  options: O | E = {} as O,
): E => {
  const metadata = globalEntityMetadata.get(Entity);
  const entity = new Entity();

  for (const column of metadata.columns) {
    (entity as any)[column.key] = parseColumn(column, entity, options);

    if ((entity as any)[column.key]) continue;

    (entity as any)[column.key] = isFunction(column.fallback)
      ? column.fallback()
      : column.fallback;
  }

  const hooks = metadata.hooks.filter((h) => h.decorator === "OnCreate");

  for (const hook of hooks) {
    hook.callback(entity);
  }

  return entity;
};
