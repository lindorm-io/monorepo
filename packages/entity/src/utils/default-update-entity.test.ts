import MockDate from "mockdate";
import {
  Column,
  Entity,
  PrimaryKeyColumn,
  UpdateDateColumn,
  VersionColumn,
} from "../decorators";
import { defaultCloneEntity } from "./default-clone-entity";
import { defaultCreateEntity } from "./default-create-entity";

const MockedDate = new Date("2024-01-01T08:00:00.000Z");
MockDate.set(MockedDate);

jest.mock("crypto", () => ({
  ...jest.requireActual("crypto"),
  randomUUID: jest
    .fn()
    .mockImplementation(() => "c8ff3952-8ba4-51a3-a4d5-2f0726c49524")
    .mockImplementationOnce(() => "e8edbe03-f52f-56c6-8d98-82e6961329fe")
    .mockImplementationOnce(() => "0c8c31df-f1cc-5381-97fc-31b3714327de")
    .mockImplementationOnce(() => "2dc92d72-9ce3-50c7-b982-59022ec393f1")
    .mockImplementationOnce(() => "88943033-fef1-5ad8-8cf8-4470be80d197")
    .mockImplementationOnce(() => "1d37a07d-4fd6-5913-85eb-2c7d721cf2a9")
    .mockImplementationOnce(() => "5f2976d2-afda-5e28-9c06-3c04865ea8df"),
}));

describe("defaultUpdateEntity", () => {
  test("should update existing entity", () => {
    @Entity()
    class TestUpdateEntity {
      @PrimaryKeyColumn()
      id!: string;

      @VersionColumn()
      version!: number;

      @UpdateDateColumn()
      updateDate!: Date;

      @Column("array", { fallback: [] })
      array!: any[];

      @Column("bigint", { fallback: BigInt(0) })
      bigint!: bigint;

      @Column("boolean", { fallback: false })
      boolean!: boolean;

      @Column("object", { fallback: {} })
      object!: Record<string, any>;
    }

    const original = defaultCreateEntity<TestUpdateEntity>(TestUpdateEntity, {
      version: 8,
      array: [1, 2, 3],
      bigint: BigInt(1),
      boolean: true,
      updateDate: new Date("2023-01-01T00:00:00.000Z"),
      object: { key: "value" },
    });

    expect(defaultCloneEntity(TestUpdateEntity, original)).toMatchSnapshot();
  });
});
