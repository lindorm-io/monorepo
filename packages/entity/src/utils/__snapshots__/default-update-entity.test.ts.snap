// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultUpdateEntity should update existing entity 1`] = `
TestUpdateEntity {
  "array": [
    1,
    2,
    3,
  ],
  "bigint": 1n,
  "boolean": true,
  "id": "0c8c31df-f1cc-5381-97fc-31b3714327de",
  "object": {
    "key": "value",
  },
  "updateDate": 2024-01-01T08:00:00.000Z,
  "version": 0,
}
`;
